### PR TITLE
Added endgame, highscores and entity factory component changes

### PR DESCRIPTION
**Pull request by Bram-Boris**
# Description
This PR adds
- Endgame screen
- Highscores screen
- Highscore write/read to json (assets/highscores.json)
- Entity Factory component refactor.
- FloatingPointComparer which should be used when comparing floats.

## Screenshot(s)

image
image
image

# How can this be tested?

1. Play a game
2. Remember if you killed people
3. Go back to the high score screen
4. See if you won.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

### CODE REVIEW

File: lib/controllers/game_controller.cpp [hunk 1]
```diff
@@ -126,9 +137,28 @@ void GameController::createGameStateManager() {
     state_systems->at(GameState::InGame)->push_back(std::make_unique<SpawnSystem>(entityManager, entityFactory));
     state_systems->at(GameState::InGame)->push_back(std::make_unique<RenderingSystem>(entityManager, *engine->getRenderer()));
 
+    //state_systems->at(GameState::EndGame)->push_back(std::make_unique<GameSystem>(entityManager, *this));
```
Comment (LuukSantegoeds) [inline lib/controllers/game_controller.cpp]: Does this need to stay?

File: lib/controllers/game_controller.cpp [hunk 2]
```diff
@@ -126,9 +137,28 @@ void GameController::createGameStateManager() {
...
+    state_systems->at(GameState::EndGame)->push_back(std::make_unique<ClickSystem>(entityManager));
+    state_systems->at(GameState::EndGame)->push_back(std::make_unique<MovementSystem>(collisionDetector, entityManager, entityFactory));
+    state_systems->at(GameState::EndGame)->push_back(std::make_unique<PhysicsSystem>(collisionDetector, entityManager, *delta_time_modifier.get()));
+    state_systems->at(GameState::EndGame)->push_back(std::make_unique<PickupSystem>(collisionDetector, entityManager, entityFactory));
+    state_systems->at(GameState::EndGame)->push_back(std::make_unique<CritterSystem>(collisionDetector, entityManager, entityFactory));
+    //state_systems->at(GameState::EndGame)->push_back(std::make_unique<WeaponSystem>(collisionDetector, entityManager, entityFactory));
```
Comment (LuukSantegoeds) [inline lib/controllers/game_controller.cpp]: And this one?

File: lib/scenes/level_scene.cpp [hunk 3]
```diff
@@ -8,9 +8,13 @@
 #include "brickengine/std/random.hpp"
 
 LevelScene::LevelScene(EntityFactory& factory, BrickEngine& engine, Json json)
-    : json(json), BeastScene<LevelScene>(factory, engine, json.getInt("width"), json.getInt("height")) {}
+    : json(json), BeastScene<LevelScene>(factory, engine, json.getInt("width"), json.getInt("height")) {
+        prepared = false;
+    }
 
 void LevelScene::performPrepare() {
+    entity_components = std::make_unique<std::vector<EntityComponents>>();;
```
Comment (LuukSantegoeds) [inline lib/scenes/level_scene.cpp]: That looks like a double ;

File: include/controllers/score_controller.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#ifndef FILE_SCORE_CONTROLLER_HPP
+#define FILE_SCORE_CONTROLLER_HPP
+
+#include "brickengine/entities/entity_manager.hpp"
+#include "data/score.hpp"
+
+class ScoreController {
```
Comment (Peter-JanGootzen) [inline include/controllers/score_controller.hpp]: This isn't really a controller. Maybe `ScoreHandler` or `ScoreIO`?

File: include/controllers/score_controller.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+public:
+    using Scores = std::unordered_map<std::string, Score>;
+    ScoreController() = default;
+    Scores readScores();
+    void writeScores(std::unordered_map<std::string, Score> scores);
```
Comment (Peter-JanGootzen) [inline include/controllers/score_controller.hpp]: These methods can be static

File: lib/scenes/end_scene.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        if(stats_component->levels_won > 0)
+            level_digits = floor(log10(stats_component->levels_won));
+        auto dots_to_add = 60 - level_digits - player_component->name.size() - text.size();
+        while(text.size() < dots_to_add) {
+            text = text + ".";
+        }
+        text = text + std::to_string(result);
+
+        auto comps = factory.createText(text, { 255, 255, 255, 255 }, 50, this->screen_width / 2, y, 750, 50, 1);
+        factory.addToEntityManager(std::move(comps), "EndScene");
```
Comment (Peter-JanGootzen) [inline lib/scenes/end_scene.cpp]: You should use `this->getTag()`

File: lib/controllers/game_controller.cpp (hunk 1 above)
Comment (MarkMeer) [reply lib/controllers/game_controller.cpp]: Nope, it will be deteled in the next commit

File: lib/controllers/game_controller.cpp (hunk 2 above)
Comment (markJavab) [reply lib/controllers/game_controller.cpp]: Nope, it will be deteled in the next commit

**Review by LuukSantegoeds — APPROVED** [review]

File: include/entities/entity_factory.hpp [hunk 7]
```diff
@@ -29,19 +32,22 @@ class EntityFactory {
...
+    EntityComponents createImage(std::string path, int x_pos, int y_pos, int x_scale, int y_scale, double relative_modifier, Layers layer, int alpha);
+    EntityComponents createPlatform(double x_pos, double y_pos, double x_scale, double y_scale, double relative_modifier, std::string path, int alpha);
+    EntityComponents createTrophy(int x, int y, int x_scale, int y_scale, double relative_modifier, Layers layer, int alpha);
+    EntityComponents createReadySign(int x, int y, int x_scale, int y_scale, double relative_modifier, Layers layer, int alpha);
+    std::vector<EntityComponents> createButton(std::string text, Color text_color, int font_size,
+        std::string texture_path, int x, int y, int x_scale, int y_scale, 
+        int alpha, double relative_modifier, std::function<void ()> on_click);
+    EntityComponents createText(std::string text, Color color, int font_size, int x, int y, int x_scale, int y_scale, double relative_modifier);
+
+    int addToEntityManager(EntityComponents entity_components, std::string scene_tag ,std::optional<std::pair<int,bool>> parent_opt = std::nullopt);
```
Comment (Peter-JanGootzen) [inline include/entities/entity_factory.hpp]: This is a quite handy function, but it should be used lightly, only when you can't access the EntityManager or got some EntityComponents from the EntityFactor. You should comment that above the function.

The function also doesn't correctly reflect the entity_manager, the scene_tag should be optional. When you are already in a scene such as in the Scene::start() or a system, you should let the entity_manager set the scene_tag for you.

File: lib/scenes/end_scene.cpp (hunk 6 above)
Comment (Peter-JanGootzen) [reply lib/scenes/end_scene.cpp]: And also check the comment on the entity_factory.hpp

File: include/systems/highscore_system.hpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "controllers/score_controller.hpp"
+#include "data/score.hpp"
+#include "systems/beast_system.hpp"
+
+class HighscoreSystem : public BeastSystem {
+public:
+    HighscoreSystem(std::shared_ptr<EntityManager> entity_manager, std::shared_ptr<EntityFactory> entity_factory, ScoreController& score_manager);
+    void update(double deltatime);
+    void createHighscores(std::string name, Score score);
```
Comment (Peter-JanGootzen) [inline include/systems/highscore_system.hpp]: This might be able to be private

**Review by Peter-JanGootzen — CHANGES_REQUESTED** [review]
:snowman:

**Review by Peter-JanGootzen — APPROVED** [review]

**Review by Peter-JanGootzen — APPROVED** [review]
🧱:man_cartwheeling:🧱

**Pull request merged**